app/scripts/views/settings/recovery_codes.js

```diff
@@ -0,0 +1,142 @@
+const ModalSettingsPanelMixin = require('../mixins/modal-settings-panel-mixin');
+const Template = require('templates/settings/recovery_codes.mustache');
+const RecoveryCodePrintTemplate = require('templates/settings/recovery_codes_print.mustache');
+const RecoveryCode = require('../../models/recovery-code');
+
+var t = FormView.t;
+
+const View = FormView.extend({
+  template: Template,
+  className: 'recovery-codes',
+  viewName: 'settings.two-step-authentication.recovery-codes',
+
+  events: {
+    'click .copy-codes': FormView.preventDefaultThen('_copyCodes'),
+    'click .download-codes': '_downloadCodes',
+    'click .print-codes': FormView.preventDefaultThen('_printCodes'),
+    'click .replace-codes-link': FormView.preventDefaultThen('_replaceRecoveryCodes'),
+    'click .two-step-authentication-done': FormView.preventDefaultThen('_returnToTwoStepAuthentication')
+  },
+
+  _returnToTwoStepAuthentication() {
+    this.navigate('settings/two_step_authentication');
+  },
+
+  _copyCodes() {
+    // This copies the recovery codes to clipboard by creating a tiny transparent
+    // textArea with recovery code contents. Then it executes the
+    // browser `copy` command.
+    $('<textArea class=\"recovery-code-text-area\"></textArea>').appendTo('#recovery-codes');
+    this.$('.recovery-code-text-area').html(this.recoveryCodesText);
+    this.$('.recovery-code-text-area').select();
+    this.$('.recovery-code-text-area').focus();
+    try {
+      this.window.document.execCommand('copy');
+      this._displaySuccess(t('Codes copied'));
+    } catch (err) {
+      this._displayError(t('Failed to copy codes'));
+    }
+    this.$('.recovery-code-text-area').remove();
+  },
+
+  _downloadCodes() {
+    // This dynamically creates a link with a blob data of the recovery
+    // codes and clicks it to initiate download.
+    const codeBlob = new Blob([this.recoveryCodesText], {type: 'text/plain'});
+    const filename = 'Recovery Codes.txt';
+    const href = URL.createObjectURL(codeBlob);
+    const template = `
+      <a id="recovery-code-download-link" href="${href}" download="${filename}"></a>
+    `;
+    $(template).appendTo('#recovery-codes');
+    this.window.document.getElementById('recovery-code-download-link').click();
+    this.$('#recovery-code-download-link').remove();
+  },
+
+  _printCodes() {
+    // We dynamically create a new window with recovery codes and attempt to
+    // print it.
+    const printWindow = this.window.open('', 'Print', 'height=600,width=800');
+    const renderCodes = this.recoveryCodes.map((code) => {
+      return new RecoveryCode({
+        code: code
+      }).toJSON();
+    });
+    const template = RecoveryCodePrintTemplate({
+      recoveryCodes: renderCodes
+    });
+    printWindow.document.write(template);
+    printWindow.document.close();
+    printWindow.focus();
+    printWindow.print();
+    printWindow.close();
+  },
+
+  _displaySuccess(msg) {
+    this.$('.error').addClass('hidden');
```

Comment:
Is displaying success and error messages this way common enough to extract this into a shared area?

Reply:
I wondered about that too, but opted not too since this was the only modal that was displaying errors.